### PR TITLE
Add support for MESSAGELIMIT Extension

### PR DESCRIPTION
**Pull request by danieleggert**
Add support for MESSAGELIMIT Extension.

### Motivation:

Some use of RFC 9394 (aka. `PARTIAL`) to some extent relies on the server being able to announce a limit on the number of messages that can be processed in a single command. That is what

https://www.ietf.org/archive/id/draft-ietf-extra-imap-messagelimit-08.html

is about.

### Modifications:

Add the `MESSAGELIMIT` Extension by adding support for
 * `MESSAGELIMIT` capability
 * `SAVELIMIT` capability
 * `UIDAFTER` search key
 * `UIDBEFORE` search key

Additionally, this PR cleans up / simplifies the `LastCommandSet` such that the `protocol IMAPEncodable` is no longer needed.

### CODE REVIEW

File: Sources/NIOIMAPCore/Grammar/Command/Command.swift [hunk 1]
```diff
@@ -94,13 +94,13 @@ public enum Command: Hashable {
     case idleStart
 
     /// Copies each message in a given set to a new mailbox, preserving the original in the current mailbox.
-    case copy(LastCommandSet<SequenceSet>, MailboxName)
+    case copy(LastCommandSet<SequenceNumber>, MailboxName)
```
Comment (Davidde94) [inline Sources/NIOIMAPCore/Grammar/Command/Command.swift:97]: @danieleggert IIRC a `SequenceSet` allows the representation `1:*`, whereas `SequenceNumber` is just the `1` or `*`. Assuming that's true, do we definitely want this change?

Comment (danieleggert) [reply Sources/NIOIMAPCore/Grammar/Command/Command.swift:97]: I changed `LastCommandSet` to be generic over the type that’s in the `MessageIdentifierSet`, and it’s thus still `1:*` here.

```
public enum LastCommandSet<N: MessageIdentifier>: Hashable {
    case set(MessageIdentifierSetNonEmpty<N>)
    …
```

Hope that makes sense?

File: EXTENSIONS.md [hunk 2]
```diff
@@ -12,22 +12,24 @@ CONDSTORE|[RFC7162](https://www.iana.org/go/rfc7162)
...
+MESSAGELIMIT|[MessageLimitDraft](https://www.ietf.org/id/draft-ietf-extra-imap-messagelimit-08.html)
 METADATA-SERVER|[RFC5464](https://www.iana.org/go/rfc5464)
+METADATA|[RFC5464](https://www.iana.org/go/rfc5464)
 MOVE|[RFC6851](https://www.iana.org/go/rfc6851)
 MULTIAPPEND|[RFC3502](https://www.iana.org/go/rfc3502)
 MULTISEARCH|[RFC7377](https://www.iana.org/go/rfc7377)
 NAMESPACE|[RFC2342](https://www.iana.org/go/rfc2342)
 PARTIAL|[RFC9394](https://www.iana.org/go/rfc9394)
 PREVIEW|[RFC8970](https://www.iana.org/go/rfc8970)
+PREVIEW|[RFC8970](https://www.iana.org/go/rfc8970)
```
Comment (mdiep) [inline EXTENSIONS.md:32]: This got duplicated

File: Sources/NIOIMAPCore/Grammar/Response/ResponseTextCode.swift [hunk 3]
```diff
@@ -273,11 +273,11 @@ public indirect enum ResponseTextCode: Hashable {
 }
 
 extension ResponseTextCode {
-    public static func modified(_ set: MessageIdentifierSet<UID>) -> Self {
+    public static func modified(_ set: MessageIdentifierSetNonEmpty<UID>) -> Self {
```
Comment (mdiep) [inline Sources/NIOIMAPCore/Grammar/Response/ResponseTextCode.swift:276]: This could be `UIDSetNonEmpty`

**Review by mdiep — APPROVED** [review]

File: Sources/NIOIMAPCore/Grammar/Command/Command.swift (hunk 1 above)
Comment (Davidde94) [reply Sources/NIOIMAPCore/Grammar/Command/Command.swift:97]: Yep makes sense!

**Pull request merged**